man/dmtxquery.1
.\" $Id$
.\"
.\" Man page for the dmtxquery utility (libdmtx project).
.\"
.\" To view: $ groff -man -T ascii dmtxquery.1 | less
.\" To text: $ groff -man -T ascii dmtxquery.1 | col -b | expand
.\"
.TH DMTXREAD 1 "September 4, 2009"
.SH NAME
dmtxquery \- extract information from dmtxread output
.SH SYNOPSIS
.B dmtxquery
\fIPROPERTY\fP [\fIOPTION\fP]... [\fIFILE\fP]...
.SH DESCRIPTION
dmtxquery extracts information from the XML output from dmtxread for individual or grouped barcode scan results.
.SH PROPERTY
.PP
barcode.count             count of all barcodes found in image
.PP
barcode.\fBN\fP.\fBBPROP\fP           print BPROP property of Nth barcode
.PP
message.count             count of all messages found in image
.PP
message.\fBN\fP.\fBMPROP\fP           print MPROP property of Nth message
.PP
message.\fBN\fP.barcode.count   count of all barcodes in Nth message
.PP
message.\fBN\fP.barcode.\fBM\fP       Mth barcode of Nth message, print all
.PP
message.\fBN\fP.barcode.\fBM\fP.\fBBPROP\fP Mth barcode of Nth message, print BPROP
.PP
\fBBPROP\fP barcode properties:
   message        message_number      message_position
   matrix_size    data_codewords      error_codewords
   rotation       data_regions_count  interleaved_blocks
.PP
\fBMPROP\fP message properties:
   message        data_codeword       error_codeword
.SH OPTIONS
.TP
\fB\-V\fP, \fB\-\-version\fP
Print program version information.
.TP
\fB\-\-help\fP
Display this help message and quit.
.SH STANDARDS
ISO/IEC 16022:2000
.PP
ANSI/AIM BC11 ISS
.SH BUGS
Email bug reports to \fImike@dragonflylogic.com\fP
.SH AUTHOR
Copyright (C) 2008, 2009 Mike Laughton
.\" end of man page


NEWS
06-Aug-2009: Simon Wood released pyDataMatrixScanner at SourceForge
03-Aug-2009: Evan Nemerson added Vala support with a new wrapper
31-Jul-2009: libdmtx Documentation wiki moved to libdmtx.wikidot.com
19-May-2009: libdmtx project discussed in Linux Journal (March 2009)
12-Mar-2009: Python wrapper supports multi-barcode scans (thanks Simon!)
02-Mar-2009: New release: libdmtx-0.7.0 (Release Notes)
31-Jan-2009: Experimental libdmtx Windows binaries now available
23-Nov-2008: New release: libdmtx-0.6.0 (Release Notes)
08-Oct-2008: Olivier Guilyardi added support for all major image formats! (SVN)
04-Sep-2008: New release: libdmtx-0.5.2 (README)
20-Aug-2008: Significant performance gains available in SVN (latest tarball)
01-Jul-2008: New release: libdmtx-0.5.1 (README)
29-Jun-2008: libdmtx.org now hosted on a super efficient fit-PC
03-Jun-2008: New iPhone reader by Stefan Hafeneger (no jailbreak required)
31-May-2008: New libdmtx mailing lists replace forums
30-May-2008: GeoWeb does it again: CameraDMTX for Nokia N60
27-May-2008: Et tu, Cognex?
22-May-2008: Congratulations Cognex! Good guys 2, patent trolls 0
16-Apr-2008: New documentation wiki. Everyone can contribute!
13-Apr-2008: 5th release: libdmtx-0.5.0 (README)
22-Feb-2008: New phone video, this time on Symbian S60
15-Jan-2008: Unofficial .debs now available (thanks pcitizen!)
21-Dec-2007: Check out Martin's iDMTX video!
07-Dec-2007: 4th release: libdmtx-0.4.0 (README)
02-Dec-2007: Good news: Project is ACTIVE - new release soon
16-Aug-2007: Added new screenshots to show progress during freeze
30-Nov-2006: Bad news: Downloads have been halted (explanation)
15-Oct-2006: 3rd release: libdmtx-0.3.0 (README)
13-Sep-2006: Unstable version now available from CVS (instructions)
07-Sep-2006: Now taking feature requests for next release
11-Jun-2006: 2nd release: libdmtx-0.2.0 (view README file before using)
22-Apr-2006: 1st release: libdmtx-0.1.0 (view README file before using)
12-Mar-2006: Project home page changed hosting
24-Jan-2006: ISO/IEC 16022:2000 specification arrived in the mail
25-Nov-2005: Project home page libdmtx.sourceforge.net created
22-Nov-2005: SourceForge project created


util/dmtxquery/dmtxquery.c
/*
libdmtx - Data Matrix Encoding/Decoding Library

Copyright (C) 2008, 2009 Mike Laughton

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

Contact: mike@dragonflylogic.com
*/

/* $Id$ */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <string.h>
#include <getopt.h>
#include <dmtx.h>
#include "dmtxquery.h"
#include "../common/dmtxutil.h"

char *programName;

/**
 * Main function for the dmtxquery Data Matrix scanning utility.
 *
 * @param argc count of arguments passed from command line
 * @param argv list of argument passed strings from command line
 * @return     numeric success / failure exit code
 */
int
main(int argc, char *argv[])
{
   int err;
   UserOptions options;

   SetOptionDefaults(&options);

   err = HandleArgs(&options, &argc, &argv);
   if(err)
      ShowUsage(err);

   exit(0);
}

/**
 *
 *
 */
static void
SetOptionDefaults(UserOptions *options)
{
   memset(options, 0x00, sizeof(UserOptions));

   /* Set default options */
}

/**
 * Sets and validates user-requested options from command line arguments.
 *
 * @param options    runtime options from defaults or command line
 * @param argcp      pointer to argument count
 * @param argvp      pointer to argument list
 * @return           DmtxPass | DmtxFail
 */
static DmtxPassFail
HandleArgs(UserOptions *options, int *argcp, char **argvp[])
{
   int opt;
   int longIndex;

   struct option longOptions[] = {
         {"version",          no_argument,       NULL, 'V'},
         {"help",             no_argument,       NULL,  0 },
         {0, 0, 0, 0}
   };

   programName = Basename((*argvp)[0]);

   if(*argcp == 1) /* Program called without arguments */
      return DmtxFail;

   for(;;) {
      opt = getopt_long(*argcp, *argvp, "V", longOptions, &longIndex);
      if(opt == -1)
         break;

      switch(opt) {
         case 0: /* --help */
            ShowUsage(0);
            break;
         case 'V':
            fprintf(stdout, "%s version %s\n", programName, DmtxVersion);
            fprintf(stdout, "libdmtx version %s\n", dmtxVersion());
            exit(0);
            break;
         default:
            return DmtxFail;
            break;
      }
   }

   return DmtxPass;
}

/**
 * Display program usage and exit with received status.
 *
 * @param status error code returned to OS
 * @return       void
 */
static void
ShowUsage(int status)
{
   if(status != 0) {
      fprintf(stderr, _("Usage: %s PROPERTY [OPTION]... [FILE]...\n"), programName);
      fprintf(stderr, _("Try `%s --help' for more information.\n"), programName);
   }
   else {
      fprintf(stdout, _("Usage: %s PROPERTY [OPTION]... [FILE]...\n"), programName);
      fprintf(stdout, _("\
Extract information from the XML output from dmtxread for individual or\n\
grouped barcode scan results.\n\
\n\
Example: dmtxread barcode.png | %s barcode.count\n\
Example: %s barcode.2.rotation scanresults.xml\n\
\n\
PROPERTY:\n"), programName, programName);
      fprintf(stdout, _("\
   barcode.count             count of all barcodes found in image\n\
   barcode.N.BPROP           print BPROP property of Nth barcode\n\
\n\
   message.count             count of all messages found in image\n\
   message.N.MPROP           print MPROP property of Nth message\n\
   message.N.barcode.count   count of all barcodes in Nth message\n\
   message.N.barcode.M       Mth barcode of Nth message, print all\n\
   message.N.barcode.M.BPROP Mth barcode of Nth message, print BPROP\n\
\n\
   BPROP barcode properties:\n\
      message        message_number      message_position\n\
      matrix_size    data_codewords      error_codewords\n\
      rotation       data_regions_count  interleaved_blocks\n\
\n\
   MPROP message properties:\n\
      message        data_codeword       error_codeword\n\
\n\
OPTIONS:\n\
  -V, --version              print program version information\n\
      --help                 display this help and exit\n"));
      fprintf(stdout, _("\nReport bugs to <mike@dragonflylogic.com>.\n"));
      fprintf(stdout, "\n\
***************************************************************\n\
**** This program currently does nothing besides printing  ****\n\
**** this help screen. This program is just a placeholder. ****\n\
***************************************************************\n\n");
   }

   exit(status);
}
